File: 156. Binary Tree Upside Down/upsideDownBinaryTree.py
from typing import Optional
# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

class Solution:
  def upsideDownBinaryTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
    if not root or not root.left:
      return root
    
    newRoot = self.upsideDownBinaryTree(root.left)
    root.left.right = root
    root.left.left = root.right
    root.left = None
    root.right = None

    return newRoot


root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)

root.left.left = TreeNode(4)
root.left.right = TreeNode(5)

solution = Solution()
print('upsideDownBinaryTree: ', solution.upsideDownBinaryTree(root))

File: 156. Binary Tree Upside Down/upsideDownBinaryTree.js
class TreeNode {
  constructor(val = 0, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

const upsideDownBinaryTree = (root) => {
  if (!root || !root.left) {
    return root;
  }

  let leftChild = root.left;
  let rightChild = root.right;

  let newRoot = upsideDownBinaryTree(leftChild);

  leftChild.right = root;
  leftChild.left = rightChild;
  root.left = null;
  root.right = null;

  return newRoot;
};

const root = new TreeNode(1)
root.left = new TreeNode(2)
root.right = new TreeNode(3)

root.left.left = new TreeNode(4)
root.left.right = new TreeNode(5)

console.log('upsideDownBinaryTree: ', upsideDownBinaryTree(root));
